edgelet/iotedge-proxy/src/error.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::fmt;
+use std::fmt::Display;
+use std::io::Error as IoError;
+
+use failure::{Backtrace, Context, Fail};
+use hyper::http::uri::InvalidUri;
+use hyper::{header, Body, Error as HyperError, Response, StatusCode};
+use native_tls::Error as NativeTlsError;
+use serde_json::json;
+use url::ParseError as UrlParseError;
+
+use crate::IntoResponse;
+
+#[derive(Debug)]
+pub struct Error {
+    inner: Context<ErrorKind>,
+}
+
+#[derive(Debug, Fail, PartialEq)]
+pub enum ErrorKind {
+    #[fail(display = "Could not load settings")]
+    LoadSettings,
+
+    #[fail(display = "Unsupported url schema {:?}", _0)]
+    UnsupportedSchema(String),
+
+    #[fail(display = "Could not initialize tokio runtime")]
+    Tokio,
+
+    #[fail(display = "Invalid URL {:?}", _0)]
+    InvalidUrl(String),
+
+    #[fail(display = "Invalid URL {:?}: {}", _0, _1)]
+    InvalidUrlWithReason(String, String),
+
+    #[fail(display = "HTTP connection error")]
+    Hyper,
+
+    #[fail(display = "A native TLS error occurred")]
+    NativeTls,
+
+    #[fail(display = "Parse error url error")]
+    Parse,
+
+    #[fail(display = "Invalid URI to parse")]
+    Uri,
+
+    #[fail(display = "Invalid HTTP header value {:?}", _0)]
+    HeaderValue(String),
+
+    #[fail(display = "An IO error occurred")]
+    Io,
+
+    #[fail(display = "An IO error occurred {:?}", _0)]
+    File(String),
+
+    #[fail(display = "Error")]
+    Generic,
+}
+
+impl Error {
+    pub fn kind(&self) -> &ErrorKind {
+        self.inner.get_context()
+    }
+}
+
+impl Fail for Error {
+    fn cause(&self) -> Option<&dyn Fail> {
+        self.inner.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.inner.backtrace()
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Display::fmt(&self.inner, f)
+    }
+}
+
+impl IntoResponse for Error {
+    fn into_response(self) -> Response<Body> {
+        let mut fail: &dyn Fail = &self;
+        let mut message = self.to_string();
+        while let Some(cause) = fail.cause() {
+            message.push_str(&format!("\n\tcaused by: {}", cause.to_string()));
+            fail = cause;
+        }
+
+        let body = json!({
+            "message": message,
+        })
+        .to_string();
+
+        Response::builder()
+            .status(StatusCode::INTERNAL_SERVER_ERROR)
```

Comment:
I wanted to return some meaningful error message when https client has an error when making a request to `iotedged`. If do nothing at all hyper simply closes the connection with HTTP client in case of error. Not sure how to improve it. Thoughts?

Reply:
I think that we should return something more meaningful via HTTP status codes. A 500 means that the server (in this case, the proxy) has failed, but in the example you gave, iotedged failed.

IMHO, I would use the following:
502 - iotedged is down
401/403 - certificate error
500 - general proxy error unrelated to iotedged

Reply:
502 and 500 - good point. 
Certificate error will be returned by iotedged, so we can pass response as is, I think